### DISCUSSION TITLE
`ee_filter`s

### DISCUSSION THREAD

**Discussion by zackarno** [joshualerickson/exploreRGEE#14] [Ideas]
I like the new `ee_filter` functions. It would be really cool to be able to plug these into all sorts of code and not need to think about mapping with `ee_utils_pyfunc`  & `ee$filter$calendarRange` anymore. One potential issue with doing this though is that it seems the final step of these functions is to composite the image with the chosen stat/reducer. Therefore, I am wondering if it might make sense to:

1. Separate the actual filter functionality into it’s own function(s). Something like `ee_filter_calendar` or even `ee_filter_calendar_range`. I feel like they could be all merged into one filter with a “by” argument (“month”,”year”,”year_month”)
2. Create a second function which does what the current `ee_filter` functions do which is filter + composite/reduce. This function would wrap the new filter function mentioned above. I am thinking something like `ee_temporal_composite` or `ee_calendar_composite`. I think this naming/verbage would be consistent with [GEE documentation on compositing](https://developers.google.com/earth-engine/guides/ic_composite_mosaic)

Below is a rough skeleton of how I am currently imagining the functions could fit together just to give you the basic idea. There is a bunch missing and I don't think it simplifies the function code a great deal, but i think it could improve the syntax and also make the filter more portable. Let me know what you think. If you like the idea, I would be happy to give it a go.

```{r}
#' @param imageCol imageCollection
#' @param by time unit ("month", "year" , "year_month")
#' @param months \code{numeric} vector, e.g. c(1,12).
#' @param years \code{numeric} vector min/max years e.g. c(2013, 2016).
#' 
ee_filter_calendar_range(imageCol, by, month, years){
  
  if(by=="months"){
    
  }
  if(by="years"){
    
  }
  if(by="year_month")
}
  
}


#' @param imageCol imageCollection
#' @param by time unit ("month", "year" , "year_month")
#' @param months \code{numeric} vector, e.g. c(1,12).
#' @param years \code{numeric} vector min/max years e.g. c(2013, 2016).
#' @param stat \code{character} statistic
 

ee_calendar_composite<- function(imageCol,by,stat,months, years){
  

  ee_reducer <- stat_to_reducer(stat)
  # filter
  imageCol_temp_filtered <- ee_calendar_filter(imageCol,by=by, months= months, years=years)
  # composite
  imageCol_temp_composite <-ee_reducer(imageCol_temp_filtered)
  
 # have to map again to set the properties - kind of annoying. I guess we could write a function to set properties based on "by", "month", 
 # and year
 
  if(by=="months"){
    months = ee$List$sequence(months[1], months[2])
    ee$ImageCollection$fromImages(
      months$map(rgee::ee_utils_pyfunc(function (m) {
        indexString = ee$Number(m)$format('%03d')
        ic_temp_filtered <-imageCol_temp_composite $
          set('system:index', indexString)$
          set('year',0000)$
          set('month',m)$
          set('date',ee$Date$fromYMD(1,m,1))$
          set('system:time_start',ee$Date$millis(ee$Date$fromYMD(1,m,1)))
      }
      )))
  }  
  if(by=="years"){
    
    # write some code
  }
  if(by=="year_month"){
    
    # some code
  }
  
    
}
```

**Comment by joshualerickson** [comment]
I'm looking at it right now. I like it. I have some questions below.

**ee_filter_calendar_range()**
So the first function `ee_filter_calendar_range()` would this have functionality like below? Where you are essentially abstracting away `ee$Filter$calendarRange()` and `imageCollection$filterDate()`? I wonder if we'd even need the `by = ` call? We could just catch whats missing or what's not missing and apply filter from there.

```
imageCol |> ee_filter_calendar_range(months = c(8,12))
imageCol |> ee_filter_calendar_range(months = c(8,12), years = c(1990, 1999))
```

**ee_calendar_composite()**

This is essentially the `ee_filter_*` but with a `by = ` argument, right? I think we could just have a top-level function `ee_calendar_composite()` and then whatever the `by` argument is that will call the `ee_filter_*` functions. I like this idea of combining, 👍. 

I'm down if you want to tackle this!

**Reply by zackarno** [reply]
Nice, I will try combining and make a version where the condition is the "by" argument and another where it is based on which inputs are filled (month vs year vs month and year). We can then test them out and see how they feel
